components/ws-manager/pkg/manager/manager.go

```diff
@@ -346,6 +361,119 @@ func (m *Manager) StartWorkspace(ctx context.Context, req *api.StartWorkspaceReq
+	err = m.Clientset.Get(ctx, types.NamespacedName{Namespace: "", Name: volumeSnapshotClassName}, &volumeSnapshotClass)
+	if err != nil {
+		return fmt.Errorf("was unable to get volume snapshot class: %v", err)
+	}
+
+	snapshotContentName := "restored-" + id
+	volumeSnapshot := &volumesnapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      id,
+			Namespace: m.Config.Namespace,
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotSpec{
+			Source: volumesnapshotv1.VolumeSnapshotSource{
+				VolumeSnapshotContentName: &snapshotContentName,
+			},
+		},
+	}
+	if volumeSnapshotClassName != "" {
+		volumeSnapshot.Spec.VolumeSnapshotClassName = &volumeSnapshotClassName
+	}
+
+	err = m.Clientset.Create(ctx, volumeSnapshot)
+	if err != nil && !k8serr.IsAlreadyExists(err) {
+		return fmt.Errorf("cannot create volumesnapshot: %v", err)
+	}
+
+	volumeSnapshotContent := &volumesnapshotv1.VolumeSnapshotContent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      snapshotContentName,
+			Namespace: "", // content is not namespaced
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotContentSpec{
+			VolumeSnapshotRef: corev1.ObjectReference{
+				Kind:      "VolumeSnapshot",
+				Name:      id,
+				Namespace: m.Config.Namespace,
+			},
+			DeletionPolicy: "Delete",
+			Source: volumesnapshotv1.VolumeSnapshotContentSource{
+				SnapshotHandle: &handle,
+			},
+			Driver: volumeSnapshotClass.Driver,
+		},
+	}
+
+	err = m.Clientset.Create(ctx, volumeSnapshotContent)
+	if err != nil && !k8serr.IsAlreadyExists(err) {
+		return fmt.Errorf("cannot create volumesnapshotcontent: %v", err)
+	}
+
+	return nil
+}
+
+func (m *Manager) DeleteVolumeSnapshot(ctx context.Context, req *api.DeleteVolumeSnapshotRequest) (res *api.DeleteVolumeSnapshotResponse, err error) {
+	span, ctx := tracing.FromContext(ctx, "DeleteVolumeSnapshot")
+	tracing.LogRequestSafe(span, req)
+	defer tracing.FinishSpan(span, &err)
+	log := log.WithField("func", "DeleteVolumeSnapshot")
+
+	okResponse := &api.DeleteVolumeSnapshotResponse{}
+
+	var volumeSnapshot volumesnapshotv1.VolumeSnapshot
+	err = m.Clientset.Get(ctx, types.NamespacedName{Namespace: m.Config.Namespace, Name: req.Id}, &volumeSnapshot)
+	if k8serr.IsNotFound(err) {
+		if !req.SoftDelete {
+			err = m.restoreVolumeSnapshotFromHandle(ctx, req.Id, req.VolumeHandle)
+			if err != nil {
+				log.WithError(err).Error("was unable to restore volume snapshot")
+				return nil, err
+			}
+		} else {
+			return okResponse, nil
+		}
+	} else if err != nil {
+		log.WithError(err).Error("was unable to get volume snapshot")
+		return nil, err
+	}
+
+	err = m.Clientset.Delete(ctx,
+		&volumesnapshotv1.VolumeSnapshot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      req.Id,
+				Namespace: m.Config.Namespace,
+			},
+		},
+	)
+	if err != nil && !k8serr.IsNotFound(err) {
+		log.WithError(err).Errorf("failed to delete volume snapshot `%s`", req.Id)
+		return nil, err
+	}
+	if !k8serr.IsNotFound(err) {
```

Comment:
Should this one be `if k8serr.IsNotFound(err) {`?

Reply:
no. we want to report back if we indeed deleted snapshot volume. if it was not found, then nothing was actually deleted.